confluent_kafka/producer.py

```diff
@@ -0,0 +1,130 @@
+
+        if not isinstance(conf, dict):
+            raise TypeError("expected configuration dict")
+
+        if kwargs:
+            # Handle kwargs for backwards compatibility
+            conf.update(kwargs)
+            warn("The use of kwargs is being deprecated. "
+                 "In future releases `conf` will be mandatory and "
+                 "all keyword arguments must match the constructor signature explicitly.",
+                 category=DeprecationWarning, stacklevel=2)
+
+        self._key_serializer = key_serializer
+        self._value_serializer = value_serializer
+
+        # Callbacks can be set in the conf dict or *ideally* as parameters.
+        # Handle both cases prior to passing along to _impl
+        # If callbacks are configured in both places parameter values take precedence.
+        if not error_cb:
+            error_cb = conf.get('error_cb', None)
+
+        if not stats_cb:
+            stats_cb = conf.get('stats_cb', None)
+
+        if not throttle_cb:
+            throttle_cb = conf.get('throttle_cb', None)
+
+        if not logger:
+            logger = conf.get('logger', None)
+
+        super(Producer, self).__init__(conf, error_cb=error_cb, stats_cb=stats_cb,
+                                       throttle_cb=throttle_cb, logger=logger)
+
+    def produce(self, topic, value=None, key=None, partition=PARTITION_UA,
+                on_delivery=None, callback=None, timestamp=0, headers=None,
+                key_serializer=None, value_serializer=None):
+        """
+            Produces message to Kafka
+
+            :param str topic: Topic to produce message to.
+            :param str|bytes|obj value: Message payload; value_serializer required for non-character values.
+            :param str|bytes|obj key: Message key payload; key_serializer required for non-character keys.
+            :param int partition: Partition to produce to, else uses the configured built-in partitioner.
+                Default value is PARTITION_UA(round-robin over all partitions).
+            :param func on_delivery(confluent_kafka.KafkaError, confluent_kafka.Message):
+                Delivery report callback to call on successful or failed delivery.
+            :param func callback(confluent_kafka.KafkaError, confluent_kafka.Message):
+                See on_delivery.
+            :param int timestamp: Message timestamp (CreateTime) in milliseconds since epoch UTC
+                (requires librdkafka >= v0.9.4, api.version.request=true, and broker >= 0.10.0.0).
+                Default value is current time.
+            :param dict|list headers: Message headers to set on the message. The header key must be a string while
+                the value must be binary, unicode or None. Accepts a list of (key,value) or a dict.
+                (Requires librdkafka >= v0.11.4 and broker version >= 0.11.0.0)
+            :param func key_serializer(topic, key): Producer key_serializer override;
```

Comment:
maybe call these `default_key_serializer` and `default_value_serializer`.

Reply:
I would like to encourage per client serializer configs despite having to support per request serializer configurations because of the existing Avro client API. With that said I'd like to keep the name `default` out of the name if possible with hope that we will find in practice noone actually wants to override their serializers per request.

Reply:
semantically they are treated as defaults, so it's less muddled to call them that. if you don't think the feature will be used, then it'd be better to exclude it than providing a sub-optimal api. but it's not un-common to want to use a single librdkafka instance to produce messages of different types. three ways to achieve this are 1. as you have it, 2. one duo of serializers per producer + ability to create other producer instances based on the same underlying librdkafka instance. 3. registering / lookup on type (or even topic).

i'd have to try and use the api to get a feel for whether specifying on each call is too cumbersome (it is in C#, but might not be in python).

Reply:
1. I think is the most flexible and how I personally would like to proceed. 

2. Sharing an underlying librdkafka insance brings about its own problems. For instance what do you do when one of the other producer instances calls a blocking operation such as `flush`?

3. I think you should register alternative strategies with your serializer instead of with the producer. This keeps the state of stateful serializer where it belongs, in the serializer. 

Realistically I don't foresee anyone overriding the serializer per call. If they did there was no reason to implement generic serdes in the first place. This could have been done with the existing interface. Instead I think the serializer should have logic built into it which handles the various states you need it to. For Avro this would mean pulling the schema from the generic record, creating a datum writer if you don't have one for this schema already and then writing the record. This is the job of the serializer though not  the producer. The producer's job should be to handle over the necessary information and take the resulting output and that's it.